src/frontc/cparser.mly

```diff
@@ -704,57 +717,93 @@ constant:
     CST_INT				{CONST_INT (fst $1), snd $1}
 |   CST_FLOAT				{CONST_FLOAT (fst $1), snd $1}
 |   CST_COMPLEX     {CONST_COMPLEX (fst $1), snd $1}
-|   CST_CHAR				{CONST_CHAR (fst $1), snd $1}
-|   CST_WCHAR				{CONST_WCHAR (fst $1), snd $1}
-|   string_constant		        {CONST_STRING (fst $1), snd $1}
-|   wstring_list			{CONST_WSTRING (fst $1), snd $1}
+|   CST_CHAR {CONST_CHAR (fst $1), snd $1}
+|   CST_WCHAR				{CONST_WCHAR (fst $1, WCHAR_T), snd $1}
+|   CST_CHAR16      {CONST_WCHAR (fst $1, CHAR16_T), snd $1}
+|   CST_CHAR32      {CONST_WCHAR (fst $1, CHAR32_T), snd $1}
+|   string_constant     {
+        let queue, typ, location = $1 in
+        match typ with
+        | CHAR -> CONST_STRING (queue_to_string queue, NO_ENCODING), location
+        | CHAR_UTF8 -> CONST_STRING (queue_to_string queue, UTF8), location
+        | _ -> CONST_WSTRING (queue_to_int64_list queue, typ), location
+    }
 ;
 
-string_constant:
-/* Now that we know this constant isn't part of a wstring, convert it
-   back to a string for easy viewing. */
-    string_list                         {
-     let queue, location = $1 in
-     let buffer = Buffer.create (Queue.length queue) in
-     Queue.iter
-       (List.iter
-	  (fun value ->
-	    let char = int64_to_char value in
-	    Buffer.add_char buffer char))
-       queue;
-     Buffer.contents buffer, location
-   }
-;
 one_string_constant:
 /* Don't concat multiple strings.  For asm templates. */
     CST_STRING                          {intlist_to_string (fst $1) }
 ;
-string_list:
+
+string_constant:
```

Comment:
Why did this have to change names?

Reply:
To use the name `string_constant` only for specific types of strings (as it was used before) is a bit confusing. So the name became free and it makes sense to use it here.

Reply:
I changed it somewhat to stay closer to the original names while still removing the confusion.